flake.nix

```diff
@@ -351,27 +401,101 @@
-              sed -i "s|solc = .*|solc = \"${solcDefault}/bin/solc\"|g" ethereum/contracts/foundry.toml
+              sed "s|# solc = .*|solc = \"${solcDefault}/bin/solc\"|g" \
+                ethereum/contracts/foundry.toml.in > \
+                ethereum/contracts/foundry.toml
+            '';
+            postShellHook = ''
+              ${pre-commit-check.shellHook}
             '';
           };
           defaultDevShell = buildDevShell [ ];
           smoketestsDevShell = buildDevShell [ hoprd hopli ];
+          update-github-labels = flake-utils.lib.mkApp {
+            drv = pkgs.writeShellScriptBin "update-github-labels" ''
+              set -eu
+              # remove existing crate entries (to remove old crates)
+              yq 'with_entries(select(.key != "crate:*"))' .github/labeler.yml > labeler.yml.new
+              # add new crate entries for known crates
+              for f in `find . -mindepth 2 -name "Cargo.toml" -type f ! -path "./vendor/*"`; do
+              	env \
+              		name="crate:`yq '.package.name' $f`" \
+              		dir="`dirname $f`/**" \
+              		yq -n '.[strenv(name)][0]."changed-files"[0]."any-glob-to-any-file"[0] = env(dir)' >> labeler.yml.new
+              done
+              mv labeler.yml.new .github/labeler.yml
+            '';
+          };
         in
         {
+          treefmt = {
+            inherit (config.flake-root) projectRootFile;
+
+            programs.yamlfmt.enable = true;
+            settings.formatter.yamlfmt.includes = [ "./.github/labeler.yml" "./.github/workflows/*.yaml" ];
+            settings.formatter.yamlfmt.excludes = [ "./vendor/*" ];
+
+            programs.prettier.enable = true;
+            settings.formatter.prettier.includes = [ "*.md" "*.json" ];
+            settings.formatter.prettier.excludes = [ "./vendor/*" "./ethereum/contracts/broadcast/*" ];
+
+            programs.rustfmt.enable = true;
+            settings.formatter.rustfmt.excludes = [ "./vendor/*" ];
+
+            programs.nixpkgs-fmt.enable = true;
+            settings.formatter.nixpkgs-fmt.excludes = [ "./vendor/*" ];
+
+            programs.ruff.enable = true;
+            programs.ruff.format = true;
+            settings.formatter.ruff.excludes = [ "./vendor/*" ];
+
+            settings.formatter.solc = {
+              command = "sh";
+              options = [
+                "-euc"
+                ''
+                  # must generate the foundry.toml here, since this step could
+                  # be executed in isolation
+                  sed "s|# solc = .*|solc = \"${solcDefault}/bin/solc\"|g" \
+                    ./ethereum/contracts/foundry.toml.in > \
+                    ./ethereum/contracts/foundry.toml
+
+                  for file in "$@"; do
+                    ${pkgs.foundry-bin}/bin/forge fmt $file \
+                      --root ./ethereum/contracts;
+                  done
+                ''
+                "--"
```

Comment:
Seeing this on so many places, could this not be a separate dependency?

Reply:
What exactly are you referring to?

Reply:
This part:
```nix
                  sed "s|# solc = .*|solc = \"${solcDefault}/bin/solc\"|g" \
                    ./ethereum/contracts/foundry.toml.in > \
                    ./ethereum/contracts/foundry.
```

Reply:
We can eventually optimize that, sure.